### PR TITLE
Added webhook package

### PR DESCRIPTION
**Pull request by evgenLevin**

### CODE REVIEW

**Review by cdvultur — COMMENTED** [review]
Overall it looks good. 

I think it requires the extra *AdditionalOptions type definition for the mutating and validating webhooks.

Also by looking at the K8s api, I think the WithWebhooks could be added at this initial phase.

File: pkg/webhook/mutatingwebhook.go [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+	if valid, err := builder.validate(); !valid {
+		return &admregv1.MutatingWebhookConfiguration{}, err
+	}
+
+	mutatingWebhookConfiguration := &admregv1.MutatingWebhookConfiguration{}
+	err := builder.apiClient.Get(context.TODO(), goclient.ObjectKey{
+		Name: builder.Definition.Name,
+	}, mutatingWebhookConfiguration)
+
+	if err != nil {
```
Comment (yprokule) [inline pkg/webhook/mutatingwebhook.go:82]: I'd log the error to make it easier to troubleshoot

File: pkg/webhook/mutatingwebhook.go [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+		return &admregv1.MutatingWebhookConfiguration{}, err
+	}
+
+	return mutatingWebhookConfiguration, err
+}
+
+// Delete removes a MutatingWebhookConfiguration from a cluster.
+func (builder *MutatingWebhookConfigurationBuilder) Delete() (*MutatingWebhookConfigurationBuilder, error) {
```
Comment (yprokule) [inline pkg/webhook/mutatingwebhook.go:90]: is `DeleteAndWait` method(like there are for other packages) relevant here?

File: pkg/webhook/validatingwebhook.go [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+	if valid, err := builder.validate(); !valid {
+		return &admregv1.ValidatingWebhookConfiguration{}, err
+	}
+
+	validatingWebhookConfiguration := &admregv1.ValidatingWebhookConfiguration{}
+	err := builder.apiClient.Get(context.TODO(), goclient.ObjectKey{
+		Name: builder.Definition.Name,
+	}, validatingWebhookConfiguration)
+
+	if err != nil {
```
Comment (yprokule) [inline pkg/webhook/validatingwebhook.go:82]: ditto, worth logging an error

File: pkg/webhook/validatingwebhook.go [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+		return &admregv1.ValidatingWebhookConfiguration{}, err
+	}
+
+	return validatingWebhookConfiguration, err
+}
+
+// Delete removes a ValidatingWebhookConfiguration from a cluster.
+func (builder *ValidatingWebhookConfigurationBuilder) Delete() (*ValidatingWebhookConfigurationBuilder, error) {
```
Comment (yprokule) [inline pkg/webhook/validatingwebhook.go:90]: ditto, what about `DeleteAndWait`

**Comment by yprokule** [comment]
Isn't `Update` method required for those 2 resources?

File: pkg/webhook/mutatingwebhook.go [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+	Object *admregv1.MutatingWebhookConfiguration
+	// Used in functions that define or mutate MutatingWebhookConfiguration definitions.
+	// errorMsg is processed before MutatingWebhookConfiguration object is created.
+	errorMsg string
+	// apiClient opens api connection to the cluster.
+	apiClient *clients.Settings
+}
+
+// PullMutatingWebhookConfiguration pulls existing MutatingWebhookConfiguration from cluster.
+func PullMutatingWebhookConfiguration(apiClient *clients.Settings, name string) (
```
Comment (kononovn) [inline pkg/webhook/mutatingwebhook.go:31]: Please remove Webhook from function name.

File: pkg/webhook/validatingwebhook.go [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+
+	admregv1 "k8s.io/api/admissionregistration/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metaV1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	goclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// ValidatingWebhookConfigurationBuilder provides struct for ValidatingWebhookConfiguration object
+// which contains connection to cluster and ValidatingWebhookConfiguration definition.
+type ValidatingWebhookConfigurationBuilder struct {
```
Comment (kononovn) [inline pkg/webhook/validatingwebhook.go:18]: Please remove webhook.

File: pkg/webhook/validatingwebhook.go [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+	Object *admregv1.ValidatingWebhookConfiguration
+	// Used in functions that define or mutate ValidatingWebhookConfiguration definitions.
+	// errorMsg is processed before ValidatingWebhookConfiguration object is created.
+	errorMsg string
+	// apiClient opens api connection to the cluster.
+	apiClient *clients.Settings
+}
+
+// PullValidatingWebhookConfiguration pulls existing ValidatingWebhookConfiguration from cluster.
+func PullValidatingWebhookConfiguration(apiClient *clients.Settings, name string) (
```
Comment (kononovn) [inline pkg/webhook/validatingwebhook.go:31]: Please remove webhook

File: pkg/webhook/validatingwebhook.go [hunk 8]
```diff
@@ -0,0 +1,136 @@
+package webhook
```
Comment (kononovn) [inline pkg/webhook/validatingwebhook.go:1]: Please add Create and Update functions

File: pkg/webhook/mutatingwebhook.go [hunk 9]
```diff
@@ -0,0 +1,136 @@
+package webhook
```
Comment (kononovn) [inline pkg/webhook/mutatingwebhook.go:1]: Please add Create and Update functions

**Review by kononovn — CHANGES_REQUESTED** [review]
Please check my comments above.

File: pkg/webhook/validatingwebhook.go (hunk 8 above)
Comment (evgenLevin) [reply pkg/webhook/validatingwebhook.go:1]: create is not relevant

**Comment by evgenLevin** [comment]
@yprokule @kononovn @cdvultur  Thanks for your comments. PTAL one more time

File: pkg/webhook/mutatingwebhook.go [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+		return builder, fmt.Errorf("can not delete MutatingWebhookConfiguration: %w", err)
+	}
+
+	builder.Object = nil
+
+	return builder, nil
+}
+
+// DeleteAndWait deletes an MutatingWebhookConfiguration and waits until it is removed from the cluster.
+func (builder *MutatingConfigurationBuilder) DeleteAndWait(timeout time.Duration) error {
```
Comment (kononovn) [inline pkg/webhook/mutatingwebhook.go:116]: I don't think DeleteAndWait is relevant for this object. It removes immediately from cluster and doesn't have statuses like "Terminating" or "Pending".

File: pkg/webhook/validatingwebhook.go [hunk 11]
```diff
@@ -0,0 +1,184 @@
...
+		return builder, fmt.Errorf("can not delete ValidatingWebhookConfiguration: %w", err)
+	}
+
+	builder.Object = nil
+
+	return builder, nil
+}
+
+// DeleteAndWait deletes an ValidatingWebhookConfiguration and waits until it is removed from the cluster.
+func (builder *ValidatingConfigurationBuilder) DeleteAndWait(timeout time.Duration) error {
```
Comment (kononovn) [inline pkg/webhook/validatingwebhook.go:117]: I don't think DeleteAndWait is relevant for this object. It removes immediately from cluster and doesn't have statuses like "Terminating" or "Pending".

**Review by kononovn — CHANGES_REQUESTED** [review]
2 more small comments.

File: pkg/webhook/mutatingwebhook.go (hunk 10 above)
Comment (evgenLevin) [reply pkg/webhook/mutatingwebhook.go]: removed

File: pkg/webhook/validatingwebhook.go (hunk 11 above)
Comment (evgenLevin) [reply pkg/webhook/validatingwebhook.go]: removed

File: pkg/webhook/mutatingwebhook.go (hunk 1 above)
Comment (evgenLevin) [reply pkg/webhook/mutatingwebhook.go:81]: added

**Review by kononovn — APPROVED** [review]
/lgtm

File: pkg/webhook/validatingwebhook.go (hunk 3 above)
Comment (evgenLevin) [reply pkg/webhook/validatingwebhook.go:82]: added

File: pkg/webhook/validatingwebhook.go (hunk 4 above)
Comment (evgenLevin) [reply pkg/webhook/validatingwebhook.go]: As Nikita menshioned no need for this resource

**Review by achuzhoy — APPROVED** [review]

**Pull request merged**